### PR TITLE
feat: implment BackendRefs in EnvoyProxy

### PR DESCRIPTION
**Pull request by zirain**
xref: https://github.com/envoyproxy/gateway/issues/3041

### CODE REVIEW

File: tools/make/golang.mk [hunk 1]
```diff
@@ -56,17 +56,20 @@ go.testdata.complete: ## Override test ouputdata
...
-		go test ./... --tags=integration,celvalidation -race -coverprofile=coverage.xml -covermode=atomic
+		go test ./... --tags=integration -race -coverprofile=coverage.xml -covermode=atomic
 
 .PHONY: go.test.cel
 go.test.cel: manifests $(tools/setup-envtest) # Run the CEL validation tests
 	@$(LOG_TARGET)
-	go clean -testcache # Ensure we're not using cached test results
-	KUBEBUILDER_ASSETS="$(shell $(tools/setup-envtest) use $(ENVTEST_K8S_VERSION) -p path)" \
-		go test ./test/cel-validation --tags=celvalidation -race
+	@for ver in $(ENVTEST_K8S_VERSIONS); do \
```
Comment (zirain) [inline tools/make/golang.mk:67]: `!has(self.backendRef) || !has(self.backendRef.kind) || self.backendRef.kind == 'Service'` passed on `1.27.1` and failed on `1.28.0`

so we need run CEL on different version of kubernetes.

File: api/v1alpha1/accesslogging_types.go [hunk 2]
```diff
@@ -96,11 +96,14 @@ type FileEnvoyProxyAccessLog struct {
 
 // OpenTelemetryEnvoyProxyAccessLog defines the OpenTelemetry access log sink.
 //
-// +kubebuilder:validation:XValidation:message="BackendRef only support Service Kind.",rule="!has(self.backendRef) || !has(self.backendRef.kind) || self.backendRef.kind == 'Service'"
+// +kubebuilder:validation:XValidation:message="host or backendRef needs to be set",rule="has(self.host) || has(self.backendRef)"
+// +kubebuilder:validation:XValidation:message="backendRef only support Service Kind.",rule="!has(self.backendRef) || self.backendRef.kind == 'Service'"
```
Comment (zirain) [inline api/v1alpha1/accesslogging_types.go:100]: TBH, the previous version was more rigorous, but fails on `1.28.0`.
Good news, there's dafault value on `kind`.

**Comment by arkodg** [comment]
blocked on https://github.com/envoyproxy/gateway/issues/3091

File: api/v1alpha1/shared_types.go [hunk 3]
```diff
@@ -407,3 +408,10 @@ type KubernetesPatchSpec struct {
 	// Object contains the raw configuration for merged object
 	Value apiextensionsv1.JSON `json:"value"`
 }
+
+// BackendRef defines how an ObjectReference that is specific to BackendRef.
+type BackendRef struct {
```
Comment (arkodg) [inline api/v1alpha1/shared_types.go:413]: 🚀

File: internal/xds/bootstrap/bootstrap.go [hunk 4]
```diff
@@ -154,15 +155,23 @@ func GetRenderedBootstrapConfig(opts *RenderBootsrapConfigOptions) (string, erro
 			}
 
 			// skip duplicate sinks
-			addr := fmt.Sprintf("%s:%d", sink.OpenTelemetry.Host, sink.OpenTelemetry.Port)
+			var host string
+			var port uint32
+			if sink.OpenTelemetry.Host != nil {
+				host, port = *sink.OpenTelemetry.Host, uint32(sink.OpenTelemetry.Port)
+			}
+			if len(sink.OpenTelemetry.BackendRefs) > 0 {
```
Comment (arkodg) [inline internal/xds/bootstrap/bootstrap.go:163]: maybe a TODO here to support multiple backendRefs in the future

File: internal/utils/net/backendref.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+package net
+
+import (
+	"fmt"
+
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+func BackendHostAndPort(backendRef gwapiv1.BackendObjectReference, defaultNamespace string) (string, uint32) {
```
Comment (arkodg) [inline internal/utils/net/backendref.go:14]: can we find a way to reuse https://github.com/envoyproxy/gateway/blob/c943042cb61f5248df6bf9f38d555d4222f65c0d/internal/gatewayapi/route.go#L1068 instead ?

File: internal/xds/bootstrap/bootstrap.go (hunk 4 above)
Comment (zirain) [reply internal/xds/bootstrap/bootstrap.go:163]: EG support multi sinks, I don't think that future exists.

File: internal/utils/net/backendref.go (hunk 5 above)
Comment (zirain) [reply internal/utils/net/backendref.go:14]: let's address this first and then fix it with follow up PR.

File: api/v1alpha1/envoyproxy_metric_types.go [hunk 6]
```diff
@@ -51,11 +49,13 @@ type ProxyMetricSink struct {
...
 //
-// +kubebuilder:validation:XValidation:message="BackendRef only support Service Kind.",rule="!has(self.backendRef) || !has(self.backendRef.kind) || self.backendRef.kind == 'Service'"
+// +kubebuilder:validation:XValidation:message="host or backendRefs needs to be set",rule="has(self.host) || self.backendRefs.size() > 0"
 type ProxyOpenTelemetrySink struct {
 	// Host define the service hostname.
 	// Deprecated: Use BackendRef instead.
-	Host string `json:"host"`
+	//
+	// +optional
+	Host *string `json:"host"`
```
Comment (arkodg) [inline api/v1alpha1/envoyproxy_metric_types.go:58]: omitempty

File: api/v1alpha1/tracing_types.go [hunk 7]
```diff
@@ -41,20 +39,24 @@ type TracingProvider struct {
 	Type TracingProviderType `json:"type"`
 	// Host define the provider service hostname.
 	// Deprecated: Use BackendRef instead.
-	Host string `json:"host"`
+	//
+	// +optional
+	Host *string `json:"host"`
```
Comment (arkodg) [inline api/v1alpha1/tracing_types.go:44]: omitempty

File: internal/utils/net/backendref.go (hunk 5 above)
Comment (davidalger) [reply internal/utils/net/backendref.go:14]: there is a working implementation using endpoints for the backendRef on the #3062 draft that I'd put together while waiting on APIs to be settled so I could test things out end to end while I had time to focus on it.

might be able to pull pieces of that and use them here, particularly the common bits in the controller that are required to watch the services and endpoints for backendRefs on EnvoyProxies

**Review by guydc — APPROVED** [review]
LGTM

**Review by davidalger — APPROVED** [review]

**Review by arkodg — APPROVED** [review]

**Pull request merged**